### PR TITLE
Fix citadel sync issue

### PR DESCRIPTION
**Pull request by clyang82**
This PR is to fix the `citadel` sync issue. sync service account depends on the secret cache, but the secret cache can be empty. Here is output of `citadel` pod:
```
2018-11-20T05:03:42.396581Z	error	Failed to create secret in attempt 1/3, (error: secrets "istio.pv-protection-controller" already exists)
2018-11-20T05:03:43.401264Z	error	Failed to create secret in attempt 2/3, (error: secrets "istio.pv-protection-controller" already exists)
2018-11-20T05:03:44.419085Z	error	Failed to create secret in attempt 3/3, (error: secrets "istio.pv-protection-controller" already exists)
2018-11-20T05:03:45.419233Z	error	Failed to create secret for service account "pv-protection-controller"  (error: secrets "istio.pv-protection-controller" already exists), retries 3 times
```
This error occurs because the secret cache is not synced.

This fix is to check secret cache is synced, then start service account  gorouter.

### CODE REVIEW

**Comment by clyang82** [comment]
@ymesika Thanks for your comments. Just updated the PR.

**Comment by clyang82** [comment]
`/assign @incfly`

File: security/pkg/k8s/controller/workloadsecret.go [hunk 1]
```diff
@@ -178,7 +178,15 @@ func NewSecretController(ca ca.CertificateAuthority, certTTL time.Duration, trus
 // Run starts the SecretController until a value is sent to stopCh.
 func (sc *SecretController) Run(stopCh chan struct{}) {
 	go sc.scrtController.Run(stopCh)
-	go sc.saController.Run(stopCh)
+	for {
```
Comment (hzxuzhonghu) [inline security/pkg/k8s/controller/workloadsecret.go]: replace with `cache.WaitForCacheSync()`

**Comment by hzxuzhonghu** [comment]
Also should update here: check if it is already exist error. 
```
		_, err = sc.core.Secrets(saNamespace).Create(secret)
		if err == nil {
			break
		} else {
			log.Errorf("Failed to create secret in attempt %v/%v, (error: %s)", i+1, secretCreationRetry, err)
		}
```

File: security/pkg/k8s/controller/workloadsecret.go (hunk 1 above)
Comment (myidpt) [reply security/pkg/k8s/controller/workloadsecret.go]: Please fix as suggested by @hzxuzhonghu

Comment (clyang82) [reply security/pkg/k8s/controller/workloadsecret.go]: @hzxuzhonghu @myidpt Thanks for your suggestion. I am working on

**Comment by clyang82** [comment]
> Also should update here: check if it is already exist error.
> 
> ```
> 		_, err = sc.core.Secrets(saNamespace).Create(secret)
> 		if err == nil {
> 			break
> 		} else {
> 			log.Errorf("Failed to create secret in attempt %v/%v, (error: %s)", i+1, secretCreationRetry, err)
> 		}
> ```

@hzxuzhonghu already checked above your attached code.
```
_, exists, err := sc.scrtStore.Get(secret)
	if err != nil {
		log.Errorf("Failed to get secret from the store (error %v)", err)
	}
```
Do we need to check it again?

**Comment by hzxuzhonghu** [comment]
>Do we need to check it again?

They are not same. I am pretty sure should do it.

**Comment by clyang82** [comment]
@hzxuzhonghu  @incfly  @myidpt  @ymesika Thanks for your comments. Please review again.

File: security/pkg/k8s/controller/workloadsecret.go [hunk 2]
```diff
@@ -255,14 +260,19 @@ func (sc *SecretController) upsertSecret(saName, saNamespace string) {
...
-		if err == nil {
+		if err == nil || errors.IsAlreadyExists(err) {
 			break
 		} else {
 			log.Errorf("Failed to create secret in attempt %v/%v, (error: %s)", i+1, secretCreationRetry, err)
 		}
 		time.Sleep(time.Second)
 	}
 
+	if err != nil && errors.IsAlreadyExists(err) {
```
Comment (ymesika) [inline security/pkg/k8s/controller/workloadsecret.go]: nit `err != nil` is redundant here

Comment (hzxuzhonghu) [inline security/pkg/k8s/controller/workloadsecret.go]: remove this

File: security/pkg/k8s/controller/workloadsecret.go [hunk 3]
```diff
@@ -255,14 +260,19 @@ func (sc *SecretController) upsertSecret(saName, saNamespace string) {
...
+		// Do nothing for the already exist error.
+		return
+	}
+
 	if err != nil {
```
Comment (hzxuzhonghu) [inline security/pkg/k8s/controller/workloadsecret.go]: replace with `if err != nil && !errors.IsAlreadyExists(err)`

Comment (clyang82) [reply security/pkg/k8s/controller/workloadsecret.go]: Do you think we should throw this info here `log.Infof("Istio secret for service account \"%s\" in namespace \"%s\" has been created", saName, saNamespace)`? if apply your suggestion, this information message shows up.

Comment (hzxuzhonghu) [reply security/pkg/k8s/controller/workloadsecret.go]: why not log this information?

File: security/pkg/k8s/controller/workloadsecret.go [hunk 4]
```diff
@@ -255,15 +260,15 @@ func (sc *SecretController) upsertSecret(saName, saNamespace string) {
 	// We retry several times when create secret to mitigate transient network failures.
 	for i := 0; i < secretCreationRetry; i++ {
 		_, err = sc.core.Secrets(saNamespace).Create(secret)
-		if err == nil {
+		if err == nil || errors.IsAlreadyExists(err) {
```
Comment (myidpt) [inline security/pkg/k8s/controller/workloadsecret.go]: nit:
Use log.Infof() in the case of errors.IsAlreadyExists(err).

Comment (clyang82) [reply security/pkg/k8s/controller/workloadsecret.go]: fixed

**Comment by clyang82** [comment]
@hzxuzhonghu & @myidpt Thanks for your patient and help. Could you please merge this PR if you have no new comments?

**Pull request merged**